### DISCUSSION TITLE
iTwin 3.0 error:  FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory

### DISCUSSION THREAD

**Discussion by JohnMajerle** [iTwin/itwinjs-core#3133] [Q&A]
Hello,

I ran into an interesting problem that I don't know how to debug.  After porting my app based on the demo-portal sample to iTwin 3.0, I get the following error when executing "yarn start"

After googling for this error I tried increasing the memory several times, eventually giving up at 16 gb:

`set NODE_OPTIONS=--max_old_space_size=16384`

...but this did not fix the problem.

Has anyone else experience node running out of heap memory?  What could cause this to happen?

Thanks,

John M.

<--- Last few GCs --->

[19680:000001BDE5E24DC0]   311864 ms: Mark-sweep 4035.3 (4140.6) -> 4021.9 (4143.1) MB, 3584.0 / 0.1 ms  (average mu = 0.256, current mu = 0.104) allocation failure scavenge might not succeed
[19680:000001BDE5E24DC0]   316941 ms: Mark-sweep 4038.1 (4143.3) -> 4023.2 (4144.6) MB, 4921.4 / 0.1 ms  (average mu = 0.141, current mu = 0.031) allocation failure scavenge might not succeed

<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF73667401F v8::internal::CodeObjectRegistry::~CodeObjectRegistry+112511
 2: 00007FF736603146 DSA_meth_get_flags+65542
 3: 00007FF736603FFD node::OnFatalError+301
 4: 00007FF736F35ADE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF736F2000D v8::SharedArrayBuffer::Externalize+781
 6: 00007FF736DC35FC v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF736DC0734 v8::internal::Heap::CollectGarbage+4244
 8: 00007FF736DBE0B0 v8::internal::Heap::AllocateExternalBackingStore+2000
 9: 00007FF736DDBD20 v8::internal::FreeListManyCached::Reset+1408
10: 00007FF736DDC3D5 v8::internal::Factory::AllocateRaw+37
11: 00007FF736DF181B v8::internal::FactoryBase<v8::internal::Factory>::NewRawOneByteString+75
12: 00007FF736DEA2DC v8::internal::Factory::NewStringFromUtf8+124
13: 00007FF736F3241A v8::String::NewFromUtf8+202
14: 00007FF736518621 v8::internal::OrderedNameDictionary::OrderedNameDictionary+34209
15: 00007FF73661E99B v8::internal::Malloced::operator delete+3659
16: 00007FF736F57B9F v8::internal::SetupIsolateDelegate::SetupHeap+53823
17: 000001BDE97F759C
error Command failed with exit code 134.

**Comment by JohnMajerle** [comment]
A quick follow up.  After posting I realized I needed to put this setting inside my packages.json instead of the environment.  Its now taking a very long time to run and I'm still waiting to see if it crashes anyway.  I'm still curious what could be the cause of the problem.  Am I loading an excessive number of npm packages?

**package.json snippet:** 

```
"scripts": {
    "start": "react-scripts --max_old_space_size=16384 start",
    "build": "react-scripts --max_old_space_size=16384 build",
    "test": "react-scripts --max_old_space_size=16384 test",
    "eject": "react-scripts eject"
  },
```

**Reply by calebmshafer** [reply]
@JohnMajerle can you try adding these two variables to your `.env` file to see if it helps?

```
# Advanced iModel.js CRA Config: https://github.com/imodeljs/create-react-app/blob/imodeljs/packages/react-scripts/README-imodeljs.md
USE_FAST_SASS=true

SKIP_PREFLIGHT_CHECK=true
```

**Reply by calebmshafer** [reply]
cc: @linasburneika

**Reply by JohnMajerle** [reply]
Hello Caleb, ( @calebmshafer )
Adding the two variables didn't fix it.  Any other suggestions?
Thanks,

**Reply by aruniverse** [reply]
try setting the following in your .env

```
# https://create-react-app.dev/docs/advanced-configuration/
# https://github.com/imodeljs/create-react-app/blob/imodeljs/packages/react-scripts/README-imodeljs.md
GENERATE_SOURCEMAP=false
TRANSPILE_DEPS=false
```

**Reply by JohnMajerle** [reply]
Hello Arun ( @aruniverse )
Those two variables didn't fix it.
By the way, I have node v16.13.2  installed.  From googling I found that some versions of node have experienced similar problems.  What version of node are you all using?

**Reply by aruniverse** [reply]
I recommend using the latest 14.x ver

**Reply by JohnMajerle** [reply]
Hello Arun ( @aruniverse ),
For what its worth, I removed node and installed v14.17.3 (LTS) and still get the error.
--John M.

**Reply by linasburneika** [reply]
The demo-portal isn't officially ported to itwinjs v3 yet, but I didn't face out of memory problems with other projects based on this version.
It might be you have infinite recursion somewhere in your code probably left by mistake when doing a port, maybe something else. Try removing components as many as possible from your app, and then bring them back incrementally to see by trial/error when the problem reappears. It might help to narrow down the codebase for analysis. 

You can also launch node with debugger attached and record heap snapshots to see what takes most of the memory. There are tutorials online explaining how to do it.

**Comment by JohnMajerle** [comment]
Another follow up.  Configured with more memory and after considerable time it eventually failed again but with a different stack during heap garbage collection.

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF785F8401F v8::internal::CodeObjectRegistry::~CodeObjectRegistry+112511
 2: 00007FF785F13146 DSA_meth_get_flags+65542
 3: 00007FF785F13FFD node::OnFatalError+301
 4: 00007FF786845ADE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF78683000D v8::SharedArrayBuffer::Externalize+781
 6: 00007FF7866D35FC v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1468
 7: 00007FF7866E02A9 v8::internal::Heap::PublishPendingAllocations+1129
 8: 00007FF7866DD27A v8::internal::Heap::PageFlagsAreConsistent+2842
 9: 00007FF7866CFEF9 v8::internal::Heap::CollectGarbage+2137
10: 00007FF7866CE0B0 v8::internal::Heap::AllocateExternalBackingStore+2000
11: 00007FF7866F29E6 v8::internal::Factory::NewFillerObject+214
12: 00007FF786425C25 v8::internal::DateCache::Weekday+1797
13: 00007FF7868D34B1 v8::internal::SetupIsolateDelegate::SetupHeap+494417
14: 000001D26F38A195
error Command failed with exit code 134.

**Comment by Carl-Hinkle** [comment]
@Josh-Schifter

**Comment by kmurari103** [comment]
Recently I updated my packages from 3.0.2 to 3.1.0 and started getting the same error for my viewer app. Is there a solution for this?
I had to switch back to 3.0.2 for my app to work fine
```
<--- Last few GCs --->

[12848:000001CDD1354010]   174233 ms: Scavenge (reduce) 4084.6 (4100.5) -> 4084.0 (4101.7) MB, 6.3 / 0.0 ms  (average mu = 0.133, current mu = 0.064) allocation failure 
[12848:000001CDD1354010]   174239 ms: Scavenge (reduce) 4084.7 (4100.7) -> 4084.2 (4101.7) MB, 5.9 / 0.0 ms  (average mu = 0.133, current mu = 0.064) allocation failure 
[12848:000001CDD1354010]   174248 ms: Scavenge (reduce) 4084.9 (4103.7) -> 4084.9 (4104.2) MB, 7.8 / 0.0 ms  (average mu = 0.133, current mu = 0.064) allocation failure 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 00007FF6F8895D2F napi_wrap+133327
 2: 00007FF6F882F606 SSL_get_quiet_shutdown+63062
 3: 00007FF6F883049D node::OnFatalError+301
 4: 00007FF6F9113CDE v8::Isolate::ReportExternalAllocationLimitReached+94
 5: 00007FF6F90F8ACD v8::SharedArrayBuffer::Externalize+781
 6: 00007FF6F8FA22DC v8::internal::Heap::EphemeronKeyWriteBarrierFromCode+1516
 7: 00007FF6F8FAD6FA v8::internal::Heap::ProtectUnprotectedMemoryChunks+1258
 8: 00007FF6F8FAA839 v8::internal::Heap::PageFlagsAreConsistent+2457
 9: 00007FF6F8F9F3D1 v8::internal::Heap::CollectGarbage+2049
10: 00007FF6F8F9D5D5 v8::internal::Heap::AllocateExternalBackingStore+1349
11: 00007FF6F8FBDA4B v8::internal::Factory::NewFillerObject+203
12: 00007FF6F8CEC291 v8::internal::interpreter::JumpTableTargetOffsets::iterator::operator=+1409
13: 00007FF6F919F10D v8::internal::SetupIsolateDelegate::SetupHeap+474253
14: 0000020D41B1F134
```

Thanks

**Comment by pmconne** [comment]
@kmurari103 is the background map turned on?
@kckst8 @aruniverse

**Reply by kmurari103** [reply]
Hi @pmconne 
Yes i do have map and terrain turned on. Here is snippet of how i'm doing that currently on 3.X

onImodelConnected i call below code- 

```
...
...
const availRealityModels = await SmartDeviceAPI.getRealityModels(vp.iModel); 
      for (const model of availRealityModels) {
        style.attachRealityModel(model);
        vp.displayStyle = style;
      }
      const displayStyle: DisplayStyle3dSettingsProps = {
        backgroundMap: {
          applyTerrain: true,
          terrainSettings: {
            providerName: "CesiumWorldTerrain",
            applyLighting: false,
            heightOriginMode: TerrainHeightOriginMode.Geodetic,
          },
        }
      };
        vp.overrideDisplayStyle(displayStyle);
...
...
```

getRealityModels() method-

```
  public static async getRealityModels(imodel: IModelConnection): Promise<ContextRealityModelProps[]> {
    const RealityDataClient = new RealityDataAccessClient();
    const available: RealityDataResponse = await RealityDataClient.getRealityDatas(await IModelApp.authorizationClient!.getAccessToken(), imodel.iTwinId, undefined);
    const availableModels: ContextRealityModelProps[] = [];

    for (const rdEntry of available.realityDatas) {
      const name = undefined !== rdEntry.displayName ? rdEntry.displayName : rdEntry.id;
      const rdSourceKey = {
        provider: RealityDataProvider.ContextShare,
        format: rdEntry.type === "OPC" ? RealityDataFormat.OPC : RealityDataFormat.ThreeDTile,
        id: rdEntry.id,
      };
      const tilesetUrl = await IModelApp.realityDataAccess?.getRealityDataUrl(imodel.iTwinId, rdSourceKey.id);
      if (tilesetUrl) {
        const entry: ContextRealityModelProps = {
          rdSourceKey,
          tilesetUrl,
          name,
          description: rdEntry?.description,
          realityDataId: rdSourceKey.id,
        };

        availableModels.push(entry);
        break;
      }
    }

    return availableModels;
  }
```

**Reply by aruniverse** [reply]
The OOM error happens when you build source maps. You can get around this by setting `GENERATE_SOURCEMAP=false` in your `.env` file. We're looking into fixing this

**Reply by aruniverse** [reply]
This should be fixed in 3.1.2, can you try it out and let us know if you're still running into issues?

**Reply by kmurari103** [reply]
Hi @aruniverse ,
I updated packages and now everything works fine.
Thanks
